lib/pbench/agent/results.py

```diff
@@ -122,45 +176,137 @@ def make_result_tb(self) -> str:
         with mdlog_name.open("w") as fp:
             mdlog.write(fp)
 
+        # FIXME: /bin/cp pbench.log ${pbench_run_name}/
+
         tarball = self.target_dir / f"{pbench_run_name}.tar.xz"
+        e_file = self.target_dir / f"{pbench_run_name}.tar.err"
+        args = [self.tar_path, "--create", "--force-local"]
+        if single_threaded:
+            args.append("--xz")
+        args.append(pbench_run_name)
         try:
-            with tarfile.open(tarball, mode="x:xz") as tar:
-                for f in self.result_dir.rglob("*"):
-                    tar.add(os.path.realpath(f))
-        except tarfile.TarError:
-            self.logger.error(
-                "Tar ball creation failed for {}, skipping", self.result_dir
-            )
+            # Invoke tar directly for efficiency.
+            if single_threaded:
+                with tarball.open("w") as ofp, e_file.open("w") as efp:
+                    tar_proc = subprocess.Popen(
+                        args,
+                        cwd=str(self.result_dir.parent),
+                        stdin=None,
+                        stdout=ofp,
+                        stderr=efp,
+                    )
```

Comment:
Python3 is not involved.  The two processes are connected by pipes and they execute.  The creating process does not get involved except to close at the end.

Reply:
It looks to me like the Python interpreter is responsible for collecting the outputs and writing them to their respective files.  I have no idea how `tar` or `xz` would do those writes, but I would expect that they are highly optimized for doing bulk output; I have no idea how Python handles collecting output from subprocesses and writing it to files, but if I had to bet, I would bet on `tar` and/or `xz` over Python.

It seems to me that we're likely to get the best performance by letting `tar` do all of the I/O, which is why I suggested using `call()` to get Python out of the way.  I think the input comes from the file system, and the output goes to the file system, and all we care about is whether the command completes successfully, so I don't see how being involved in the plumbing (especially if it means that `stdout` is a "stream" instead of a block device) is going to help.

Reply:
The pipes are connected together.  Python3 is not involved.

Reply:
I'm willing to believe what you say (I'm having trouble wrapping my head completely around this at the moment), but I still think there is a difference.

I think the code as it stood is in effect running a command like `tar cf - > tarball.tar` which I _think_ produces a subtly different result from `tar cf tarball.tar`.  Now, it's quite possible that I'm way out in the weeds on this one, but in the second case, `tar` does the file open and is fully aware of the nature of the file it is writing to (including its device characteristics), while in the first case, the output is a shell redirect, and I'm not sure that `tar` knows anything about it other than that it's a byte stream.  If I'm right about this, then, in the second case, `tar` can use I/O optimizations which aren't available in the first case.